feat: add C implementation for `stats/base/dists/geometric/median`

0PrashantYadav0 opened the pull request:
Resolves #3636

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for `@stdlib/stats/base/dists/geometric/median` along with relevant tests, benchmarks and examples.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3636 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C implementation for geometric median

0PrashantYadav0 committed:
fix: replace name of mean to median

0PrashantYadav0 committed:
fix: added constrain on p

0PrashantYadav0 committed:
chore: completed the code review task

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/median/lib/native.js:
```diff
@@ -0,0 +1,57 @@
+/**
+* Returns the median of a geometric distribution.
+*
+* @param {Probability} p - success probability
```
```suggestion
*
* @private
* @param {Probability} p - success probability
```

0PrashantYadav0 committed:
Merge branch 'develop' into geometric/median

Signed-off-by: JoyBoy <144602492+0PrashantYadav0@users.noreply.github.com>

0PrashantYadav0 commented:
/stdlib merge

0PrashantYadav0 committed:
fix: changes in benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/median/src/main.c:
```diff
@@ -0,0 +1,42 @@
+	) {
+		return 0.0 / 0.0; // NaN
+	}
+	return ceil( -1.0 / ( log2( 1.0 - p ) ) ) - 1.0;
```
We should use the stdlib equivalents for `ceil` and `log2` here instead of the ones from `math.h`.

Planeshifter reviewed (changes requested):
Thanks for the PR!

In the implementation, we should use the respective stdlib packages instead of those from `math.h`.

0PrashantYadav0 committed:
chore: changes as per code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
@Planeshifter made all the changes please review and merge this pr.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/median/src/main.c:
```diff
@@ -0,0 +1,44 @@
+#include "stdlib/stats/base/dists/geometric/median.h"
+#include "stdlib/math/base/assert/is_nan.h"
+#include "stdlib/math/base/special/ceil.h"
+#include "stdlib/math/base/special/log2.h"
+#include <math.h>
```
```suggestion
#include "stdlib/math/base/special/log2.h"
```

Planeshifter committed:
chore: remove unneeded header

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter reviewed (dismissed):
Lgtm.

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/geometric/median/README.md:
```diff
@@ -122,6 +122,95 @@ for ( i = 0; i < 10; i++ ) {
+
+</section>
+
+<!-- /.examples -->
```
```suggestion
<!-- /.examples -->

</section>

<!-- /.c -->
```

Planeshifter committed:
docs: add missing closing C section

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
